Prims Algorithm Tutorial

Infonioknight opened the pull request:
## Change Summary
- Included an overview, the complexities of the algorithm
- Explained one leetcode problem in depth (I couldn't find too many and I wanted to leave some for the suggested problems that were simple and similar)
- Included two suggested problems.
- Highlighted any references used.

## Checklist

**If you haven't fulfilled the below requirements or even delete the entire checklist, your PR won't be reviewed and will be closed without notice. Regular contributors (with 10+ PRs) can skip this part.**

### General 

- [/ ] This Pull Request is all my own work. (You'll be blacklisted if you are caught for plagiarism.)
- [ /] I've read [CONTRIBUTING.md](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md)
- [ /] I've applied LaTex for all variables, formulas and time / space complexity instead of using backticks
- [/ ] I've started the app locally and verified all the content and all links (if applicable) are accessible correctly
- [ /] I've included Complexity Analysis (Time Complexity & Space Complexity).
- [ /] I've written my explanation well and it is easy to understand for beginners

### Tutorial

- [ /] I've read and followed the [Tutorial Template](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md#tutorial-template)
- [ x] I've explained my topic well with 2 - 3 LC problems and no external problems are used. 
- [/ ] I've provided the full working solutions to the problems used in this tutorial.
- [ /] I've provided suggested problems at the end with the given format. See [here](https://raw.githubusercontent.com/wingkwong/leetcode-the-hard-way/main/tutorials/math/number-theory/binary-exponentiation.md) as an example. If the target solution is not available, leave `solutionLink` blank.
- [/ ] I've given credits / references if I use external resources. (For an image, give credit under it. Otherwise, add a new section called References at the end (after Suggested Problems).)

### Solutions

- [x ] I've read and followed the [Solution Template](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md#solution-template)
- [ /] I've formatted my code well with [K&R Coding style](https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr)
- [ /] I've confirmed that comments are put above each line rather than writing on the same line.
- [ /] I've included a meaningful approach name for my solution. e.g. `## Approach 1: Two Pointers`.

Infonioknight committed:
Prims Algorithm Tutorial

- Included an overview, the complexities of the algorithm
- Explained one leetcode problem in depth (I couldn't find too many and I wanted to leave some for the suggested problems that were simple and similar)
- Included two suggested problems.
- Highlighted any references used.

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
```diff
@@ -9,4 +9,130 @@ keywords:
+**Note:** The above link requires the Leetcode Subscription. To view the problem, use [this](https://leetcode.ca/all/1135.html) alternate link.
+
+### Instructions: 
+- There are N cities numbered from 1 to N.
+- You are given connections, where each connections[i] = [city1, city2, cost] represents the cost to connect city1 and city2 together. The connections are all bidirectional.
+- Return the minimum cost so that for every pair of cities, there exists a path of connections (possibly of length 1) that connects those two cities together.  The cost is the sum of the connection costs used. If the task is impossible, return -1.
+
+Example: 
+**Input:** N = 3, connections = [[1,2,5],[1,3,6],[2,3,1]]
+**Output:** 6
```
1. please format with LaTex syntax. 
2. please check in local dev.  like `Example: Input: N = 3, connections = [[1,2,5],[1,3,6],[2,3,1]] Output: 6` renders in one line.

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
````diff
@@ -9,4 +9,130 @@ keywords:
+```
+Where each tuple consists of '(cost, destination)'.
+
+Now, to find the minimum cost, we use the below code.
````
use the custom code block.  search `<Tabs>` in  kruskals page for reference.

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
````diff
@@ -9,4 +9,130 @@ keywords:
+Where each tuple consists of '(cost, destination)'.
+
+Now, to find the minimum cost, we use the below code.
+```python3
````
1. let's put the step by step explanations first and show the full code at the end.
2. use the custom code block.  search `<Tabs>` in  kruskals page for reference.

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
````diff
@@ -9,4 +9,130 @@ keywords:
+- If the node hasn't been visited, we then add the cost to our solution and mark the node as visited.
+
+```python3
+for connection in adjacency_list[current]:
+        if connection[1] not in seen:
+            heapq.heappush(connection_queue, connection)
````
check indentation

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
```diff
@@ -9,4 +9,130 @@ keywords:
+        if connection[1] not in seen:
+            heapq.heappush(connection_queue, connection)
+
+if len(seen) < N: print(-1)
+else: print(solution)
```
use `return` instead of `print`.

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
```diff
@@ -9,4 +9,130 @@ keywords:
+
+export const suggestedProblems = [ { "problemName": "1584 - Min Cost to Connect All Points", "difficulty": "Medium", "leetCodeLink": "https://leetcode.com/problems/min-cost-to-connect-all-points/", "solutionLink": "" }, { "problemName": "0787 - Cheapest Flights Within K Stops", "difficulty": "Medium", "leetCodeLink": "https://leetcode.com/problems/cheapest-flights-within-k-stops/", "solutionLink" ...
+
+## References:
+1. For the Prim's Algorithm image: https://miro.medium.com/max/700/1*7kpPIPcmXr38Juh0umM6fA.jpeg
```
for images, u can simply add it under below the image.

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
```diff
@@ -9,4 +9,130 @@ keywords:
+import heapq
+
+# Number of nodes in the given tree
+N = len(adjacency_list)
```
prefer to use `n` instead of `N`

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
```diff
@@ -9,4 +9,130 @@ keywords:
+# A list which we will use to implement a priority queue
+connection_queue = []
+# Total minimum cost to traverse all the cities
+solution = 0
```
prefer to use `res` instead of `solution`

Infonioknight committed:
Update prims-algorithm.md

Incorporated the requested changes.

Infonioknight commented:
Incorporated the requested changes. Do let me know if they're okay.

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
```diff
@@ -1,12 +1,186 @@
+            heapq.heappush(connection_queue, connection)
+
+if len(seen) < n: return -1
+else: return res
```
`else:` is not necessary here since you've returned something if the condition is true.

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
````diff
@@ -1,12 +1,186 @@
+<SolutionAuthor name="@Infonioknight"/>
+  
+```py
+import random
+import heapq
+
+def connecting_cities_with_minimum_cost(adjacency_list):
+    # Number of nodes in the given tree
+    n = len(adjacency_list)
+    # A list we will use to implement a priority queue
+    connection_queue = []
+    # Total minimum cost to traverse all the cities
+    res = 0
+    # Taking the assumption that the n nodes in the adjacency list are from 1 to n
+    start = random.randint(1, n)
+    # To ensure we don't create closed loops
+    seen = set([start])
+
+    for connection in adjacency_list[start]:
+        heapq.heappush(connection_queue, connection)
+
+    while len(seen) < n and connection_queue:
+        cost, current = heapq.heappop(connection_queue)
+        if current in seen:
+            continue
+
+        res += cost
+        seen.add(current)
+
+        for connection in adjacency_list[current]:
+            if connection[1] not in seen:
+                heapq.heappush(connection_queue, connection)
+
+    if len(seen) < n: return -1
+    else: return res
````
use LC format so that i can copy the code and run directly

wingkwong replied:
same as above comment

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
```diff
@@ -1,12 +1,186 @@
+    "problemName": "1584 - Min Cost to Connect All Points", 
+    "difficulty": "Medium", 
+    "leetCodeLink": "https://leetcode.com/problems/min-cost-to-connect-all-points/", 
+    "solutionLink": "" 
```
add the solution link. we have such solution post.

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
```diff
@@ -1,12 +1,186 @@
+    "problemName": "0787 - Cheapest Flights Within K Stops", 
+    "difficulty": "Medium", 
+    "leetCodeLink": "https://leetcode.com/problems/cheapest-flights-within-k-stops/", 
+    "solutionLink": "" 
```
add the solution link. we have such solution post.

wingkwong reviewed (changes requested):
the structure is fine. please add one more example with the same structure.

Infonioknight committed:
Update prims-algorithm.md

- Added a second example problem for the same.

Infonioknight commented:
Added a second example problem, also highlighted the solutions for the problems that had one.

Infonioknight committed:
Update prims-algorithm.md

Made a few modifications to the code, indentation to make it copy-pastable in leetcode

Infonioknight commented:
Hello! I just wanted to follow up on the PR.

wingkwong commented:
hey next time please click re-request review when it's ready.

![image](https://github.com/user-attachments/assets/1024fc16-89a1-4910-983c-bbe190633449)

wingkwong reviewed (changes requested):
please verify locally once first.

<img width="1467" alt="image" src="https://github.com/user-attachments/assets/6cca4407-e2ec-4e89-8654-d71eaa8ffbef">

Infonioknight committed:
Update prims-algorithm.md

Made necessary changes to run locally

Infonioknight commented:
Apologies for the delay, was a bit caught up with work. Made the changes and tested it locally. Seems to be working fine now.

Infonioknight commented:
Hi there! Just wanted to follow up on the PR!

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
```diff
@@ -1,12 +1,370 @@
+    "problemName": "0787 - Cheapest Flights Within K Stops", 
+    "difficulty": "Medium", 
+    "leetCodeLink": "https://leetcode.com/problems/cheapest-flights-within-k-stops/", 
+    "solutionLink": "../../solutions/0700-0799/0787-cheapest-flights-within-k-stops-medium.md" 
```
broken link

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
```diff
@@ -1,12 +1,370 @@
+First, we convert the given 'connections' array, into an adjacency list for ease of traversing.
+<Tabs>
+
+<TabItem value="py1" label="Python">
```
value should be just `py`.

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
```diff
@@ -1,12 +1,370 @@
+## Example 1: [1135 - Connecting Cities With Minimum Cost](https://leetcode.com/problems/connecting-cities-with-minimum-cost/description/?envType=problem-list-v2&envId=minimum-spanning-tree)
+**Note:** The above link requires the Leetcode Subscription. To view the problem, use [this](https://leetcode.ca/all/1135.html) alternate link.
+
+### Instructions: 
+- There are N cities numbered from 1 to N.
+- You are given connections, where each `connections[i] = [city1, city2, cost]` represents the cost to connect `city1` and `city2 ` together. The connections are all bidirectional.
+- Return the minimum cost so that for every pair of cities, there exists a path of connections (possibly of length 1) that connects those two cities together.  The cost is the sum of the connection costs used. If the task is impossible, return $-1$.
+
+#### Example: 
+
+**Input:** N = `3`, connections = `[[1,2,5],[1,3,6],[2,3,1]]`
+
+**Output:** `6`
```
please check the format of other tutorials

wingkwong reviewed (changes requested):
- please apply LaTex instead of using backticks.
- for the content, i will review in depth in weekend.

Infonioknight committed:
Update prims-algorithm.md

- Changed all backtick highlighting to LaTeX
- Modified the introduction to the problems (made it an explanation instead of just listing input and output)
- Changed value for TabItems to 'py'

Infonioknight commented:
Made the requested changes. My apologies for not checking with the format of things. I think it should be compliant now.

Infonioknight commented:
Hi, just wanted to follow up on the pull request.

wingkwong commented:
@Infonioknight hey sorry for the late update. been a bit busy these months. will try to review on weekend. thanks.

Infonioknight commented:
@wingkwong no worries at all! Do take care.

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
```diff
@@ -1,12 +1,378 @@
-Coming Soon
\ No newline at end of file
+## Suggested Pre-requisites
+Before moving on to learn the Prims algorithm, it is suggested that you know what a [Minimum Spanning Tree (MST)](../graph-theory/minimum-spanning-tree.md) is.
```
can you also work on MST page?

Infonioknight replied:
Will try in a bit once work eases up a little. Hope that's okay.

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
```diff
@@ -1,12 +1,378 @@
+**Note:** Since the starting vertex is chosen at random, it is possible to have different edges included in the MST for the same graph, but the total edge weight of the MST will still have the same minimum value.
+
+### Complexity and Use-Cases
+When implemented using a priority queue (usually a binary heap), the time complexity of Prim’s algorithm is $O(E + V log V)$, where $E$ is the number of edges and $V$ is the number of vertices. This makes Prim's algorithm highly efficient for **dense** graphs, where the number of edges is close to the maximum possible. For sparse graphs (a graph with a relatively small number of edges compared to ...
```
may elaborate how to get $O(E + V log V)$

wingkwong replied:
> This makes Prim's algorithm highly efficient for **dense** graphs,

may explain why it is efficient for dense graphs but not sparse ones.

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
```diff
@@ -1,12 +1,378 @@
+
+![Prims Algorithm Image](https://miro.medium.com/max/700/1*7kpPIPcmXr38Juh0umM6fA.jpeg)
+
+_Source: https://miro.medium.com/max/700/1*7kpPIPcmXr38Juh0umM6fA.jpeg_
```
refer to the post title instead.

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
````diff
@@ -1,12 +1,378 @@
+2: [(5, 1), (1, 3)]
+3: [(6, 1), (1, 2)]
+```
+Where each tuple consists of '(cost, destination)'.
````
`'(cost, destination)'` -> `$(cost, destination)$`

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
````diff
@@ -1,12 +1,378 @@
+<SolutionAuthor name="@Infonioknight"/>
+  
+```py
+    for connection in adjacency_list[current]:
+        if connection[1] not in seen:
+            heapq.heappush(connection_queue, connection)
````
fix indentation

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
```diff
@@ -1,12 +1,378 @@
+import heapq
+
+class Solution:
+  def connecting_cities_with_minimum_cost(self, adjacency_list):
```
pls check the format in your localhost

<img width="887" alt="image" src="https://github.com/user-attachments/assets/984cda00-eb8b-443c-99b2-391e0f8f181e" />

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
```diff
@@ -1,12 +1,378 @@
+
+## Example 2: [1584 - Min Cost to Connect All Points](https://leetcode.com/problems/min-cost-to-connect-all-points)
+### Instructions: 
+You are given an array $points$ representing integer coordinates of some points on a 2D-plane, where $points[i]$ = $[xi, yi]$.
```
`xi` -> `x_i`
`yi` -> `y_i`

apply to `j` as well.

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
```diff
@@ -1,12 +1,378 @@
+adjacency_list = {}
+for i in range(len(points)):
+    current = tuple(points[i])
+    for j in range(i+1, len(points)):
```
`i+1` -> `i + 1`

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
```diff
@@ -1,12 +1,378 @@
+points = [[0,0],[2,2],[3,10],[5,2],[7,0]]
+
+adjacency_list = {}
+for i in range(len(points)):
```
`len(points)` -> `n` since it will be used more than one time

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
````diff
@@ -1,12 +1,378 @@
+<SolutionAuthor name="@Infonioknight"/>
+  
+```py
+    for connection in adjacency_list[start]:
+        heapq.heappush(connection_queue, connection)
+
+    while len(seen) < n and connection_queue:
+        cost, current = heapq.heappop(connection_queue)
+        if current in seen:
+            continue
+
+        res += cost
+        seen.add(current)
+
+        for connection in adjacency_list[current]:
+            if connection[1] not in seen:
+                heapq.heappush(connection_queue, connection)
+
+    return res
````
fix indentation

wingkwong commented on tutorials/graph-theory/prims-algorithm.md:
```diff
@@ -1,12 +1,378 @@
+        # To ensure we don't create closed loops
+        seen = set([start])
+
+        for connection in adjacency_list[start]:
```
include comments for the below code like L314 - L323. (apply to other code block)

wingkwong reviewed (changes requested):
The subheader for examples should be consistent. Also subheader shouldn't include `:`.

<img width="283" alt="image" src="https://github.com/user-attachments/assets/06f59b82-d1f5-4f51-b109-be88d801fd81" />

Infonioknight commented:
Hey, had a hectic last couple of months. Will go through everything this weekend and make all the necessary changes

Infonioknight committed:
Update prims-algorithm.md

Infonioknight committed:
Update prims-algorithm.md

Infonioknight commented:
Hello @wingkwong, I hope you're doing well. Just remembered this after a while. If you still do need this, do let me know if there are any other modifications I need to make to make it alright to merge!

Infonioknight committed:
Merge branch 'main' into patch-1
